cob_generic_can/common/src/SocketCan.cpp

```diff
@@ -90,114 +110,125 @@ void SocketCan::init()
+    }
+    else
+    {
+        std::cout << "No message recieved: " << std::endl;
+    }
+    return bRet;
 }
 
 //-------------------------------------------
 bool SocketCan::receiveMsgRetry(CanMsg* pCMsg, int iNrOfRetry)
 {
-    if (m_bInitialized == false) return false;
+    if (m_bInitialized == false)
+    {
+        return false;
+    }
 
-    // wait until msg in buffer
+    can::Frame frame;
     bool bRet = false;
-    int i=0;
+    int i = 0;
+
     do
     {
-        if (recived) {
-            pCMsg->setID(recived_frame.id);
-            pCMsg->setLength(recived_frame.dlc);
-            pCMsg->set(recived_frame.data[0], recived_frame.data[1], recived_frame.data[2], recived_frame.data[3],
-                            recived_frame.data[4], recived_frame.data[5], recived_frame.data[6], recived_frame.data[7]);
+      std::cout << "BufferedReader read: " << m_reader.read(&frame, boost::chrono::seconds(1)) << std::endl;
+        if (recived_frame.size())
+        {
+            frame = recived_frame.front();
+            recived_frame.pop_front();
+            pCMsg->setID(frame.id);
+            pCMsg->setLength(frame.dlc);
+            pCMsg->set(frame.data[0], frame.data[1], frame.data[2], frame.data[3],
+                       frame.data[4], frame.data[5], frame.data[6], frame.data[7]);
             bRet = true;
             break;
         }
         i++;
-        usleep(10000);
+        usleep(1000);
     }
-    while(i < iNrOfRetry);
+    while ((i < iNrOfRetry && bRet != true));
     return bRet;
 }
 
 //-------------------------------------------
 bool SocketCan::receiveMsgTimeout(CanMsg* pCMsg, int nSecTimeout)
 {
-    if (m_bInitialized == false) return false;
+    if (m_bInitialized == false)
+    {
+        return false;
+    }
 
-    // wait until msg in buffer
     bool bRet = false;
+    can::Frame frame;
+    usleep(nSecTimeout / 1000);
 
-    usleep(nSecTimeout/1000);
-    if (recived) {
-        pCMsg->setID(recived_frame.id);
-        pCMsg->setLength(recived_frame.dlc);
-        pCMsg->set(recived_frame.data[0], recived_frame.data[1], recived_frame.data[2], recived_frame.data[3],
-                        recived_frame.data[4], recived_frame.data[5], recived_frame.data[6], recived_frame.data[7]);
+    if (recived_frame.size() > 0)
+    {
+        frame = recived_frame.front();
+        recived_frame.pop_front();
+        pCMsg->setID(frame.id);
+        pCMsg->setLength(frame.dlc);
+        pCMsg->set(frame.data[0], frame.data[1], frame.data[2], frame.data[3], frame.data[4], frame.data[5], frame.data[6], frame.data[7]);
         bRet = true;
     }
 
     return bRet;
 }
 
-bool SocketCan::initCAN() {
-	m_bInitialized = true;
-        bool bRet = true;
-        return true;
+bool SocketCan::initCAN()
+{
+    m_bInitialized = true;
+    bool bRet = true;
+    return true;
 }
 
-void SocketCan::recive_frame(const can::Frame& frame){
+void SocketCan::recive_frame(const can::Frame& frame)
+{
 
-    if(frame.is_error){
```

Comment:
If you want to print message while debugging you could include https://github.com/ros-industrial/ros_canopen/blob/indigo-devel/socketcan_interface/include/socketcan_interface/string.h, it contains the ostream operators.